include/picongpu/fields/MaxwellSolver/ArbitraryOrderFDTD/Derivative.hpp

```diff
@@ -0,0 +1,234 @@
+
+#include <pmacc/math/Vector.hpp>
+#include <pmacc/meta/accessors/Identity.hpp>
+
+#include <cstdint>
+
+
+namespace picongpu
+{
+namespace fields
+{
+namespace maxwellSolver
+{
+namespace aoFDTD
+{
+namespace detail
+{
+    /** Abstraction of the arbitrary-order finite-difference time domain
+     *  derivative functor.
+     *
+     * @tparam T_lowerNeighbors Number of neighbors required in negative
+     *                          direction to calculate field derivative
+     *
+     *
+     * @tparam T_upperNeighbors Number of neighbors required in positive
+     *                          direction to calculate field derivative
+     *
+     * @tparam T_neighbors Number of neighbors used to calculate
+     *                     the derivative from finite differences.
+     *                     Order of derivative approximation is
+     *                     2 * T_neighbors
+     *
+     * @tparam T_direction Direction to take derivative in, 0 = x, 1 = y, 2 = z
+     */
+    template<
+        uint32_t T_lowerNeighbors,
+        uint32_t T_upperNeighbors,
+        uint32_t T_neighbors,
+        uint32_t T_direction
+     >
+    struct GeneralAofdtdDerivative
+    {
+        //! Lower margin
+        using LowerMargin = typename pmacc::math::CT::mul<
+            typename pmacc::math::CT::make_Int<
+                simDim,
+                T_lowerNeighbors
+            >::type,
+            typename pmacc::math::CT::make_BasisVector<
+                simDim,
+                T_direction,
+                int
+            >::type
+        >::type;
+
+        //! Upper margin
+        using UpperMargin = typename pmacc::math::CT::mul<
+            typename pmacc::math::CT::make_Int<
+                simDim,
+                T_upperNeighbors
+            >::type,
+            typename pmacc::math::CT::make_BasisVector<
+                simDim,
+                T_direction,
+                int
+            >::type
+        >::type;
+
+        /** Return derivative value at the given point
+         *
+         * @tparam T_DataBox data box type with field data
+         * @param data position in the data box to compute derivative at
+         */
+        template< typename T_DataBox >
+        HDINLINE typename T_DataBox::ValueType operator()(
+            T_DataBox const & data ) const
+        {
+            // Define shorthand type to access DataBox
+            using IndexType = pmacc::DataSpace< simDim >;
+
+            // Define indice vectors for data access
+            auto lowerIndex = IndexType{}; // Vector initialized with zeros
+            auto upperIndex = IndexType{};
+
+            // lowerIndex: 0 if ( Forward ) else -1
+            lowerIndex[ T_direction ] = static_cast<int32_t>(T_upperNeighbors) - static_cast<int32_t>(T_neighbors);
+            // upperIndex: 1 if ( Forward ) else 0
+            upperIndex[ T_direction ] = static_cast<int32_t>(T_neighbors) - static_cast<int32_t>(T_lowerNeighbors);
```

Comment:
Spacing is weird on these lines: no space after `<` and `(`

Reply:
Yes, everywheere like `static_cast< int32_t >( T_neighbors )`, this is the style that the rest of your code follows, and rightfully so